File: steelechallenges/trees/Breadth:Depth First Searches/challenge.js
class TreeNode {
    constructor(value){
        this.value = value;
        this.children = [];
    }
}

class Tree {
    constructor(){
        this.root = null;
    }

    traverseBFS(){
        //if there is no root, return false
        if(!this.root){
            return false
        }
        //start a new Queue
        const queue = new Queue()
        //keep a tally of all values in the tree
        const treeValues = [];
        //add root to queue
        queue.enqueue(this.root)
        //while queue is not empty
        while(queue.size !== 0){
            //get TreeNode Children
            const nodeChildren = queue.first.value.children
            //if node has children, loop and add each to queue
            if(nodeChildren.length !== 0){
                nodeChildren.forEach((child) => queue.enqueue(child));
            }
            //push the first item in the queue to the tree values
            treeValues.push(queue.first.value);
            //remove first node from queue
            queue.dequeue();
        }
        //return values, should be all TreeNodes
        return treeValues;
    }

    traverseDFS(){
         //if there is no root, return false
        if(!this.root){
            return false
        }
        const treeValues = [];
        let current = this.root;
        
        const preOrderHelper = (node) =>{
            treeValues.push(node.value);
            if(node.children.length !== 0){
                node.children.forEach((child) => {preOrderHelper(child)})
            }
            return true;
        }

        const postOrderHelper = (node) =>{
            if (node.children.length !== 0) {
              node.children.forEach(child => {
                preOrderHelper(child);
              });
            }
            treeValues.push(node.value);
            return true;
        }

        const inOrderHelper = (node) =>{
            if(node.children.length !== 0) {
                if(node.children.length > 1){
                    const halfway = Math.floor(node.children.length/2)
                    for(let i = 0; i < halfway; i++){
                        inOrderHelper(node.children[i])
                        //treeValues.push(node.children[i]);
                    }
                    treeValues.push(node.value)
                    for(let i = halfway; i < node.children.length; i++){
                        inOrderHelper(node.children[i]);
                        //treeValues.push(node.children[i]);
                    }

                } else{
                    node.children.forEach(child => {
                      inOrderHelper(child);
                    });
                    treeValues.push(node.value);
                }
            } else{
                treeValues.push(node.value);
            }
            return true;
        }
        
        inOrderHelper(current);
        return treeValues
    }

}

class QueueNode {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.first = null;
    this.last = null;
    this.size = 0;
  }
  //newnode goes to back of the line/end of the queue
  enqueue(value) {
    const newNode = new QueueNode(value);
    //if queue is empty
    if (this.size === 0) {
      this.first = newNode;
      this.last = newNode;
      // add current first pointer to new first(new node), and make new node new first
    } else {
      this.last.next = newNode;
      this.last = newNode;
    }
    //add 1 to size
    this.size++;

    return this;
  }
  // dequeue nodes off the front of the line
  dequeue() {
    //if queue is empty return false
    if (this.size === 0) return false;
    //get dequeuedNode
    const dequeuedNode = this.first;
    //get new first (could be NULL if stack is length 1)
    const newFirst = this.first.next;
    //if newFirst is null, reassign last to newFirst(null)
    if (!newFirst) {
      this.last = newFirst;
    }
    //assign new first
    this.first = newFirst;
    //remove refernce to list
    dequeuedNode.next = null;
    //remove 1 from size
    this.size--;
    //return dequeuednode
    return dequeuedNode;
  }

  log() {
    let currentNode = this.first;
    let i = 0;
    while (currentNode) {
      console.log(i, currentNode.value);
      i++;
      currentNode = currentNode.next;
    }
  }
}

const testTree = new Tree();

testTree.root = new TreeNode("H");
testTree.root.children.push(new TreeNode("e"));
testTree.root.children.push(new TreeNode("l"));
testTree.root.children[0].children.push(new TreeNode('l'));
testTree.root.children[0].children.push(new TreeNode('o'));
testTree.root.children[0].children.push(new TreeNode('W'));
testTree.root.children[1].children.push(new TreeNode('o'));
testTree.root.children[1].children.push(new TreeNode('r'));
testTree.root.children[1].children.push(new TreeNode('l'));
testTree.root.children[1].children.push(new TreeNode('d'));

//console.log(testTree.traverseBFS());
console.log(testTree.traverseDFS());

File: steelechallenges/trees/Breadth:Depth First Searches/challenge.rb

class QueueNode
    attr_accessor :value, :next
    def initialize(value)
        @value = value
        @next = nil
    end
end

class Queue 
    attr_accessor :size, :first, :last

    def initialize()
      @first = nil
      @last = nil 
      @size = 0 
    end

    def enqueue(val)
        new_node = QueueNode.new(val)

        if(self.size == 0)
            self.first = new_node
            self.last = new_node
        else
            self.last.next = new_node
            self.last = new_node
        end

        self.size +=1

        return self
    end

    # take off first node of list
    def dequeue
        if(self.size == 0)
            return false
        end
        dequeued_node = self.first
        new_first = dequeued_node.next
        if(!new_first)
            self.last = new_first
        end
        dequeued_node.next = nil
        self.first = new_first
        self.size -= 1

        return dequeued_node
    end
end

class TreeNode
    attr_accessor :value, :children
    def initialize(value)
        @value = value
        @children = []
    end
end 

class Tree
    attr_accessor :root

    def initialize()
        @root = nil 
    end

    def traverseBFS()
        if(!self.root)
            return false
        end
        line = Queue.new()
        tree_values = []
        line.enqueue(self.root)
        while(line.size != 0)
          node_children = line.first.value.children
          if(node_children.length != 0)
            node_children.each do |child|
            line.enqueue(child)
            end
          end
          tree_values.push(line.first.value.value)
          line.dequeue
        end
        return tree_values
    end

    def traverseDFS(type)
        if(!self.root)
            return false
        end
        @tree_values = []
        current = self.root

        def pre_order_helper(node)
            @tree_values.push(node.value)

            if(node.children.length != 0)
              node.children.each do |child|
                pre_order_helper(child)
              end
            end
            return true
        end
        
        def post_order_helper(node)
            if(node.children.length != 0)
              node.children.each do |child|
                post_order_helper(child)
              end
            end
            @tree_values.push(node.value)
            return true
        end

        def in_order_helper(node)
            if(node.children.length != 0)
              halfway = (node.children.length / 2).floor
              for i in (0...halfway) do 
                in_order_helper(node.children[i])
              end

              @tree_values.push(node.value)

              for i in (halfway...node.children.length) do 
                in_order_helper(node.children[i])
              end
            else
              @tree_values.push(node.value)
            end
            return true
        end

        case type 
        when "pre"
            pre_order_helper(current)
        when "post"
            post_order_helper(current)
        else
            in_order_helper(current)
        end

        @tree_values
    end
end





test_tree = Tree.new()
test_tree.root = TreeNode.new("H")
test_tree.root.children.push(TreeNode.new("e"))
test_tree.root.children.push(TreeNode.new("l"))
test_tree.root.children[0].children.push(TreeNode.new("l"))
test_tree.root.children[0].children.push(TreeNode.new("o"))
test_tree.root.children[0].children.push(TreeNode.new("W"))
test_tree.root.children[1].children.push(TreeNode.new("o"))
test_tree.root.children[1].children.push(TreeNode.new("r"))
test_tree.root.children[1].children.push(TreeNode.new("l"))
test_tree.root.children[1].children.push(TreeNode.new("d"))

#puts test_tree.traverseBFS()

test_tree2 = Tree.new()
test_tree2.root = TreeNode.new(10)
test_tree2.root.children.push(TreeNode.new(6))
test_tree2.root.children.push(TreeNode.new(15))
test_tree2.root.children[0].children.push(TreeNode.new(3))
test_tree2.root.children[0].children.push(TreeNode.new(8))
test_tree2.root.children[0].children.push(TreeNode.new(7))
test_tree2.root.children[1].children.push(TreeNode.new(20))

puts test_tree2.traverseDFS("in")

